File: python/src/a2a_utils/client/a2a_agents.py
"""Agent management for A2A applications."""

import asyncio
import json
from pathlib import Path
from typing import Any, Literal
from urllib.parse import urlparse

import httpx
from a2a.client import A2ACardResolver
from a2a.types import AgentCard
from loguru import logger

from ..types import AgentURLAndCustomHeaders


class A2AAgents:
    """Manages A2A agent cards keyed by user-defined agent IDs.

    Agents are configured via a dict mapping agent_id to config, or a JSON file path.
    """

    def __init__(
        self,
        agents: dict[str, dict[str, Any]] | str | Path | None = None,
        *,
        timeout: float = 15.0,
    ) -> None:
        """Initialize the agent manager.

        Args:
            agents: Agent config as:
              - dict: {"agent-id": {"url": "https://...", "custom_headers": {"X-API-Key": "..."}}}
              - str/Path: path to agents.json file with the same structure
              - None: empty, add agents later
            timeout: HTTP timeout in seconds for fetching agent cards (default: 15s).
        """
        self._timeout = timeout
        self._config = self._load_config(agents)
        self._agents: dict[str, AgentURLAndCustomHeaders] = {}
        self._init_errors: dict[str, str] = {}
        self._initialized: bool = False
        self._init_lock: asyncio.Lock = asyncio.Lock()

    def _load_config(
        self, agents: dict[str, dict[str, Any]] | str | Path | None
    ) -> dict[str, dict[str, Any]]:
        """Load and validate agent config."""
        if agents is None:
            return {}

        if isinstance(agents, (str, Path)):
            path = Path(agents)
            with open(path) as f:
                config: dict[str, dict[str, Any]] = json.load(f)
            return config

        return dict(agents)

    async def _ensure_initialized(self) -> None:
        """Lazily fetch all agent cards on first use (double-check locking)."""
        if self._initialized:
            return
        async with self._init_lock:
            if self._initialized:
                return

            if not self._config:
                logger.warning("No agents configured")
                self._initialized = True
                return

            tasks = [self._fetch_agent(agent_id, cfg) for agent_id, cfg in self._config.items()]
            results = await asyncio.gather(*tasks, return_exceptions=True)

            self._init_errors = {}
            for i, result in enumerate(results):
                if isinstance(result, Exception):
                    aid = list(self._config.keys())[i]
                    url = self._config[aid].get("url", "unknown")
                    error_msg = f"{type(result).__name__}: {result}"
                    self._init_errors[aid] = error_msg
                    logger.error(f"Error loading agent '{aid}' from {url}: {error_msg}")

            if self._agents:
                logger.success(f"Successfully initialized {len(self._agents)} agent(s)")
            else:
                logger.warning("No agents were successfully initialized")

            self._initialized = True

    async def _fetch_agent(self, agent_id: str, config: dict[str, Any]) -> None:
        """Fetch a single agent card and register it."""
        url = config["url"]
        custom_headers = config.get("custom_headers", {})

        base_url, card_path = self._parse_agent_card_url(url)
        async with httpx.AsyncClient(timeout=httpx.Timeout(self._timeout)) as httpx_client:
            resolver = A2ACardResolver(
                httpx_client=httpx_client,
                base_url=base_url,
                agent_card_path=card_path,
            )
            http_kwargs = {"headers": custom_headers} if custom_headers else None
            agent_card = await resolver.get_agent_card(http_kwargs=http_kwargs)

        self._agents[agent_id] = AgentURLAndCustomHeaders(
            agent_card=agent_card,
            custom_headers=custom_headers,
        )

    def _parse_agent_card_url(self, url: str) -> tuple[str, str]:
        """Parse agent card URL into base_url and card_path."""
        parsed = urlparse(url)
        base_url = f"{parsed.scheme}://{parsed.netloc}"
        card_path = parsed.path
        return base_url, card_path

    async def add_agent(
        self, agent_id: str, url: str, custom_headers: dict[str, str] | None = None
    ) -> None:
        """Register a new agent at runtime.

        Args:
            agent_id: User-defined agent identifier.
            url: Agent card URL.
            custom_headers: Optional custom HTTP headers.

        Raises:
            ValueError: If agent_id is already registered.
        """
        await self._ensure_initialized()
        if agent_id in self._agents:
            raise ValueError(f"Agent '{agent_id}' is already registered")
        config: dict[str, Any] = {"url": url, "custom_headers": custom_headers or {}}
        await self._fetch_agent(agent_id, config)

    async def get_agent(self, agent_id: str) -> AgentURLAndCustomHeaders | None:
        """Retrieve agent by ID.

        Args:
            agent_id: User-defined agent identifier.

        Returns:
            AgentURLAndCustomHeaders, or None if not found.
        """
        await self._ensure_initialized()
        return self._agents.get(agent_id)

    @property
    def initialization_errors(self) -> dict[str, str]:
        """Get errors from the most recent initialization attempt.

        Returns:
            Dict mapping agent_id to error message for agents that failed to load.
        """
        return dict(self._init_errors)

    async def get_agents(self) -> dict[str, AgentURLAndCustomHeaders]:
        """Get all registered agents.

        Returns:
            Dict mapping agent_id to AgentURLAndCustomHeaders.
        """
        await self._ensure_initialized()
        return dict(self._agents)

    def _format_agent_for_llm(
        self, card: AgentCard, detail: Literal["name", "basic", "skills", "full"]
    ) -> dict[str, Any]:
        """Format a single agent card into a summary dict.

        Args:
            card: The agent card to format.
            detail: Detail level — "name", "basic", "skills", or "full".

        Returns:
            Summary dict for the agent.
        """
        if detail == "name":
            return {"name": card.name}
        elif detail == "basic":
            return {"name": card.name, "description": card.description}
        elif detail == "skills":
            skill_names = [s.name for s in card.skills] if card.skills else []
            return {
                "name": card.name,
                "description": card.description,
                "skills": skill_names,
            }
        else:  # "full"
            skills = []
            if card.skills:
                for s in card.skills:
                    skills.append({"name": s.name, "description": s.description or ""})
            return {
                "name": card.name,
                "description": card.description,
                "skills": skills,
            }

    async def get_agent_for_llm(
        self,
        agent_id: str,
        detail: Literal["name", "basic", "skills", "full"] = "basic",
    ) -> dict[str, Any] | None:
        """Generate summary for a single agent.

        Args:
            agent_id: User-defined agent identifier.
            detail: Detail level — "name", "basic", "skills", or "full".

        Returns:
            Summary dict for the agent, or None if not found.
        """
        await self._ensure_initialized()
        agent = self._agents.get(agent_id)
        if agent is None:
            return None
        return self._format_agent_for_llm(agent.agent_card, detail)

    async def get_agents_for_llm(
        self, detail: Literal["name", "basic", "skills", "full"] = "basic"
    ) -> dict[str, dict[str, Any]]:
        """Generate summary of all agents.

        Args:
            detail: Detail level — "name", "basic", "skills", or "full".

        Returns:
            Dict mapping agent_id to summary dict, sorted by agent_id.
        """
        await self._ensure_initialized()
        return {
            agent_id: self._format_agent_for_llm(self._agents[agent_id].agent_card, detail)
            for agent_id in sorted(self._agents)
        }

File: javascript/src/client/a2a-agents.ts
// SPDX-FileCopyrightText: 2025-present A2A Net <hello@a2anet.com>
//
// SPDX-License-Identifier: Apache-2.0

/**
 * Agent management for A2A applications.
 */

import * as fs from "node:fs";
import type { AgentCard } from "@a2a-js/sdk";
import { DefaultAgentCardResolver } from "@a2a-js/sdk/client";
import type { AgentURLAndCustomHeaders } from "../types.js";

/**
 * Manages A2A agent cards keyed by user-defined agent IDs.
 *
 * Agents are configured via a dict mapping agent_id to config, or a JSON file path.
 */
export class A2AAgents {
    private config: Record<string, Record<string, unknown>>;
    private agents: Record<string, AgentURLAndCustomHeaders> = {};
    private initErrors: Record<string, string> = {};
    private initialized = false;
    private initPromise: Promise<void> | null = null;
    private readonly timeout: number;

    /**
     * Initialize the agent manager.
     *
     * @param agents - Agent config as:
     *   - object: {"agent-id": {"url": "https://...", "custom_headers": {"X-API-Key": "..."}}}
     *   - string: path to agents.json file with the same structure
     *   - null: empty, add agents later
     * @param opts.timeout - HTTP timeout in seconds for fetching agent cards (default: 15s).
     */
    constructor(
        agents: Record<string, Record<string, unknown>> | string | null = null,
        opts?: { timeout?: number },
    ) {
        this.timeout = opts?.timeout ?? 15.0;
        this.config = this.loadConfig(agents);
    }

    /** Load and validate agent config. */
    private loadConfig(
        agents: Record<string, Record<string, unknown>> | string | null,
    ): Record<string, Record<string, unknown>> {
        if (agents === null) {
            return {};
        }

        if (typeof agents === "string") {
            const text = fs.readFileSync(agents, "utf-8");
            return JSON.parse(text) as Record<string, Record<string, unknown>>;
        }

        return { ...agents };
    }

    /** Lazily fetch all agent cards on first use (double-check locking). */
    private async ensureInitialized(): Promise<void> {
        if (this.initialized) return;

        // Single-flight pattern
        if (this.initPromise === null) {
            this.initPromise = this.doInitialize();
        }
        await this.initPromise;
    }

    private async doInitialize(): Promise<void> {
        if (this.initialized) return;

        if (Object.keys(this.config).length === 0) {
            console.warn("No agents configured");
            this.initialized = true;
            return;
        }

        const entries = Object.entries(this.config);
        const results = await Promise.allSettled(
            entries.map(([agentId, cfg]) => this.fetchAgent(agentId, cfg)),
        );

        this.initErrors = {};
        for (let i = 0; i < results.length; i++) {
            const result = results[i];
            if (result.status === "rejected") {
                const aid = entries[i][0];
                const url = this.config[aid]?.url ?? "unknown";
                const reason = result.reason;
                const errorMsg = `${(reason as Error)?.constructor?.name ?? "Error"}: ${reason}`;
                this.initErrors[aid] = errorMsg;
                console.error(`Error loading agent '${aid}' from ${url}: ${errorMsg}`);
            }
        }

        if (Object.keys(this.agents).length > 0) {
            console.info(`Successfully initialized ${Object.keys(this.agents).length} agent(s)`);
        } else {
            console.warn("No agents were successfully initialized");
        }

        this.initialized = true;
    }

    /** Fetch a single agent card and register it. */
    private async fetchAgent(agentId: string, config: Record<string, unknown>): Promise<void> {
        const url = config.url as string;
        const customHeaders = (config.custom_headers as Record<string, string>) ?? {};

        const [baseUrl, cardPath] = this.parseAgentCardUrl(url);
        const resolver = new DefaultAgentCardResolver({
            fetchImpl: this.createFetchImpl(customHeaders),
        });
        const agentCard = await resolver.resolve(baseUrl, cardPath);

        this.agents[agentId] = {
            agentCard,
            customHeaders,
        };
    }

    /** Create a fetch implementation that applies the configured timeout and custom headers. */
    private createFetchImpl(customHeaders: Record<string, string>): typeof fetch {
        const wrappedFetch: typeof fetch = (input, init?) => {
            const timeoutSignal = AbortSignal.timeout(this.timeout * 1000);
            const signal =
                init?.signal != null
                    ? AbortSignal.any([init.signal, timeoutSignal])
                    : timeoutSignal;
            const mergedInit = {
                ...init,
                signal,
                headers: {
                    ...(init?.headers as Record<string, string> | undefined),
                    ...customHeaders,
                },
            };
            return fetch(input, mergedInit);
        };
        wrappedFetch.preconnect = fetch.preconnect;
        return wrappedFetch;
    }

    /** Parse agent card URL into base_url and card_path. */
    private parseAgentCardUrl(url: string): [string, string] {
        const parsed = new URL(url);
        const baseUrl = `${parsed.protocol}//${parsed.host}`;
        const cardPath = parsed.pathname;
        return [baseUrl, cardPath];
    }

    /**
     * Register a new agent at runtime.
     *
     * @param agentId - User-defined agent identifier.
     * @param url - Agent card URL.
     * @param customHeaders - Optional custom HTTP headers.
     *
     * @throws Error if agentId is already registered.
     */
    async addAgent(
        agentId: string,
        url: string,
        customHeaders?: Record<string, string>,
    ): Promise<void> {
        await this.ensureInitialized();
        if (agentId in this.agents) {
            throw new Error(`Agent '${agentId}' is already registered`);
        }
        const config: Record<string, unknown> = {
            url,
            custom_headers: customHeaders ?? {},
        };
        await this.fetchAgent(agentId, config);
    }

    /**
     * Retrieve agent by ID.
     *
     * @param agentId - User-defined agent identifier.
     *
     * @returns AgentURLAndCustomHeaders, or null if not found.
     */
    async getAgent(agentId: string): Promise<AgentURLAndCustomHeaders | null> {
        await this.ensureInitialized();
        return this.agents[agentId] ?? null;
    }

    /**
     * Get errors from the most recent initialization attempt.
     *
     * @returns Object mapping agent_id to error message for agents that failed to load.
     */
    get initializationErrors(): Record<string, string> {
        return { ...this.initErrors };
    }

    /**
     * Get all registered agents.
     *
     * @returns Object mapping agent_id to AgentURLAndCustomHeaders.
     */
    async getAgents(): Promise<Record<string, AgentURLAndCustomHeaders>> {
        await this.ensureInitialized();
        return { ...this.agents };
    }

    /**
     * Format a single agent card into a summary object.
     *
     * @param card - The agent card to format.
     * @param detail - Detail level — "name", "basic", "skills", or "full".
     *
     * @returns Summary object for the agent.
     */
    private formatAgentForLlm(
        card: AgentCard,
        detail: "name" | "basic" | "skills" | "full",
    ): Record<string, unknown> {
        if (detail === "name") {
            return { name: card.name };
        }
        if (detail === "basic") {
            return { name: card.name, description: card.description };
        }
        if (detail === "skills") {
            const skillNames = card.skills ? card.skills.map((s) => s.name) : [];
            return {
                name: card.name,
                description: card.description,
                skills: skillNames,
            };
        }
        // "full"
        const skills: Record<string, unknown>[] = [];
        if (card.skills) {
            for (const s of card.skills) {
                skills.push({ name: s.name, description: s.description ?? "" });
            }
        }
        return {
            name: card.name,
            description: card.description,
            skills,
        };
    }

    /**
     * Generate summary for a single agent.
     *
     * @param agentId - User-defined agent identifier.
     * @param detail - Detail level — "name", "basic", "skills", or "full".
     *
     * @returns Summary object for the agent, or null if not found.
     */
    async getAgentForLlm(
        agentId: string,
        detail: "name" | "basic" | "skills" | "full" = "basic",
    ): Promise<Record<string, unknown> | null> {
        await this.ensureInitialized();
        const agent = this.agents[agentId];
        if (!agent) {
            return null;
        }
        return this.formatAgentForLlm(agent.agentCard, detail);
    }

    /**
     * Generate summary of all agents.
     *
     * @param detail - Detail level — "name", "basic", "skills", or "full".
     *
     * @returns Object mapping agent_id to summary object, sorted by agent_id.
     */
    async getAgentsForLlm(
        detail: "name" | "basic" | "skills" | "full" = "basic",
    ): Promise<Record<string, Record<string, unknown>>> {
        await this.ensureInitialized();
        const result: Record<string, Record<string, unknown>> = {};
        const sortedKeys = Object.keys(this.agents).sort();
        for (const agentId of sortedKeys) {
            result[agentId] = this.formatAgentForLlm(this.agents[agentId].agentCard, detail);
        }
        return result;
    }
}
